revert back --skip-tls flag changes with deprecation notice

laxmikantbpandhare opened the pull request:
**Description of the change:**
Reverted changes for the `--skip-tls` flag and made sure that `--skip-tls-verify` marked true if the user enters `--skip-tls` with deprecation notice.


**Motivation for the change:**
We kept the `skip-tls` flag with operator SDK so that the user can use it and get notified via deprecation notice.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

laxmikantbpandhare commented:
@jmrodri - I added code change with deprecation notice. Do we need to add notification on the website docs as well?

jmrodri reviewed (changes requested):
I would not have added SkipTLS as parameters to LoadBundle or anywhere that far. What I think we should do is have the flag be on the CLI (marked as deprecated). When the user passes that flag in we should, as early as possible, capture that they set it. THEN set SkipTLSVerify = true and UseHTTP = true. 

The code after the CLI parsing should only ever care about SkipTLSVerify and UseHTTP. It should no longer know what SkipTLS means. 

The reason SkipTLS was removed from OPM was because if it was true, it also reverted to using http. They did not want it to behave that way. That's why it was split into two flags. So in order to simulate what SkipTLS used to do is to set both of the new flags to true.

jmrodri commented on changelog/fragments/revert-skiptls-changes-deprc-notice.yaml:
```diff
@@ -0,0 +1,5 @@
+entries:
+  - description: >
+      Reverted changes for `--skip-tls` flag and made sure that `--skip-tls-verify` marked true if user enters `--skip-tls` with deprecation notice.
```
```suggestion
      Reverted removal of  `--skip-tls` flag and mark the flag as deprecated.
```

laxmikantbpandhare replied:
Modified.

jmrodri commented on internal/cmd/operator-sdk/scorecard/cmd.go:
```diff
@@ -273,5 +273,5 @@ func extractBundleImage(bundleImage string) (string, error) {
 	}
 	// FEAT: enable explicit local image extraction.
-	return registryutil.ExtractBundleImage(context.TODO(), logger, bundleImage, false, false, false)
+	return registryutil.ExtractBundleImage(context.TODO(), logger, bundleImage, false, false, false, false)
```
I would not add SkipTLS to this call.

jmrodri commented on internal/olm/operator/bundle/install.go:
```diff
@@ -72,7 +72,7 @@ func (i *Install) setup(ctx context.Context) error {
 
 	// Load bundle labels and set label-dependent values.
-	labels, bundle, err := operator.LoadBundle(ctx, i.BundleImage, i.SkipTLSVerify, i.UseHTTP)
+	labels, bundle, err := operator.LoadBundle(ctx, i.BundleImage, i.SkipTLS, i.SkipTLSVerify, i.UseHTTP)
```
I would not pass SkipTLS to this call. Instead I would pass in true for both SkipTLSVerify AND UseHTTP *IF* SkipTLS was set.

laxmikantbpandhare replied:
Removed from it.

jmrodri commented on internal/olm/operator/bundleupgrade/upgrade.go:
```diff
@@ -69,7 +69,7 @@ func (u *Upgrade) setup(ctx context.Context) error {
 	}
 
-	labels, bundle, err := operator.LoadBundle(ctx, u.BundleImage, u.SkipTLSVerify, u.UseHTTP)
+	labels, bundle, err := operator.LoadBundle(ctx, u.BundleImage, u.SkipTLS, u.SkipTLSVerify, u.UseHTTP)
```
do not pass SkipTLS to this

laxmikantbpandhare replied:
Removed it.

jmrodri commented on internal/olm/operator/helpers.go:
```diff
@@ -34,8 +34,8 @@ func CatalogNameForPackage(pkg string) string {
 // LoadBundle returns metadata and manifests from within bundleImage.
-func LoadBundle(ctx context.Context, bundleImage string, skipTLSVerify bool, useHTTP bool) (registryutil.Labels, *apimanifests.Bundle, error) {
-	bundlePath, err := registryutil.ExtractBundleImage(ctx, nil, bundleImage, false, skipTLSVerify, useHTTP)
+func LoadBundle(ctx context.Context, bundleImage string, skipTLS bool, skipTLSVerify bool, useHTTP bool) (registryutil.Labels, *apimanifests.Bundle, error) {
```
no need for skipTLS on this call.

laxmikantbpandhare replied:
Removed it.

jmrodri commented on internal/registry/image.go:
```diff
@@ -50,6 +50,11 @@ func ExtractBundleImage(ctx context.Context, logger *log.Entry, image string, lo
 	logger = logger.WithFields(log.Fields{"dir": bundleDir})
 
+	//if user set --skip-tls then set --skip-tls-verify to true as --skip-tls is deprecated
+	if skipTLS {
+		skipTLSVerify = true
+	}
+
```
This check should be done much earlier in the process, closer to where cobra is parsing the flag.

laxmikantbpandhare replied:
Removed it.

asmacdo commented:
If we release without this, it would have a backwards incompatible change, so we need to block the 1.19 release until this is merged. Please ping @theishshah when this merges.

jmrodri commented:
> Sure, I updated the code and made sure that LoadBundle will not have SkipTLS flag as a parameter. One quick question, I tried to set both skip-tls-verify and use-http flags but run bundle is failing with below error Error: invalid flag combination: --use-http and --skip-tls-verify cannot both be true. That's why I set only skip-tls-verify to true if the user enters skip-tls flag thru the command line. Does this change sound good?
@laxmikantbpandhare I'll look at what those flags mean on the `opm` side.

laxmikantbpandhare commented:
> > Sure, I updated the code and made sure that LoadBundle will not have SkipTLS flag as a parameter. One quick question, I tried to set both skip-tls-verify and use-http flags but run bundle is failing with below error Error: invalid flag combination: --use-http and --skip-tls-verify cannot both be true. That's why I set only skip-tls-verify to true if the user enters skip-tls flag thru the command line. Does this change sound good?
> 
> @laxmikantbpandhare I'll look at what those flags mean on the `opm` side.
> 
> Okay, `opm` uses `use-http` when `skip-tls` is being passed in. So that's what we want to do instead. https://github.com/operator-framework/operator-registry/blob/master/cmd/opm/internal/util/util.go#L34-L37

That's great. Modified the code accordingly.

jmrodri commented on internal/olm/operator/bundle/install.go:
```diff
@@ -71,6 +71,11 @@ func (i *Install) setup(ctx context.Context) error {
 		}
 	}
 
+	//if user set --skip-tls then set --skip-tls-verify to true as --skip-tls is deprecated
```
I think comment is now out of date. 
```suggestion
	/ /if user sets --skip-tls then set --use-http to true as --skip-tls is deprecated
```

laxmikantbpandhare replied:
Updated

jmrodri commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -77,6 +77,9 @@ type RegistryPod struct { //nolint:maligned
 	// The secret's key for this file must be "cert.pem".
 	CASecretName string
 
+	// SkipTLS controls wether to ignore SSL errors while pulling bundle image from registry server.
```
```suggestion
	// SkipTLS controls whether to ignore SSL errors while pulling bundle image from registry server.
```

laxmikantbpandhare replied:
Updated.

jmrodri reviewed (changes requested):
a couple typos in the comments.

laxmikantbpandhare commented:
> a couple typos in the comments.

Updated typos.

jmrodri reviewed (approved):
/lgtm

rashmigottipati reviewed (changes requested):
`skipTLS` field is added to registry pod struct but is not being used in the file. It should be handled in the the container creation command.

rashmigottipati commented:
You need to add `--skip-tls={{ $.SkipTLS }}` to `opm registry add` command during container creation.

rashmigottipati commented:
/hold

laxmikantbpandhare commented:
> You need to add `--skip-tls={{ $.SkipTLS }}` to `opm registry add` command during container creation.

@rashmigottipati - Is there any need for this as we are setting `--use-http` flag once we encounter `--skip-tls` flag.

rashmigottipati commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -77,6 +77,9 @@ type RegistryPod struct { //nolint:maligned
 	// The secret's key for this file must be "cert.pem".
 	CASecretName string
 
+	// SkipTLS controls whether to ignore SSL errors while pulling bundle image from registry server.
```
```suggestion
```

Since this is not being used in registry pod creation anymore.

laxmikantbpandhare replied:
removed this comment.

rashmigottipati commented:
I looked into the latest opm registry add command and skipTLS flag is removed. Since we are defaulting to use `--use-http` flag when we encounter --skip-tls and also marking it as deprecated, I'm good with the changes. 

/hold remove

rashmigottipati commented on internal/olm/operator/registry/index/registry_pod.go:
```diff
@@ -77,6 +77,8 @@ type RegistryPod struct { //nolint:maligned
 	// The secret's key for this file must be "cert.pem".
 	CASecretName string
 
+	SkipTLS bool `json:"SkipTLS"`
```
```suggestion
```

laxmikantbpandhare replied:
removed it.

rashmigottipati reviewed (approved):
/lgtm

laxmikantbpandhare commented:
Removed `release-blocker` tag as it got merged now.
